crates/nu-cli/src/shell/filesystem_shell.rs

```diff
@@ -1002,6 +663,59 @@ impl Shell for FilesystemShell {
     }
 }
 
+struct TaggedPathBuf<'a>(&'a PathBuf, &'a Tag);
+
+fn move_file(from: TaggedPathBuf, to: TaggedPathBuf) -> Result<(), ShellError> {
+    let TaggedPathBuf(from, from_tag) = from;
+    let TaggedPathBuf(to, to_tag) = to;
+
+    if to.exists() && from.is_dir() && to.is_file() {
+        return Err(ShellError::labeled_error(
+            "Cannot rename a directory to a file",
+            "invalid destination",
+            to_tag,
+        ));
+    }
+
+    let destination_dir_exists = if to.is_dir() {
+        true
+    } else {
+        to.parent().map(Path::exists).unwrap_or(true)
+    };
+
+    if !destination_dir_exists {
+        return Err(ShellError::labeled_error(
+            "Destination directory does not exist",
+            "destination does not exist",
+            to_tag,
+        ));
+    }
+
+    let mut to = to.clone();
+    if to.is_dir() {
+        let from_file_name = match from.file_name() {
+            Some(name) => name,
+            None => {
+                return Err(ShellError::labeled_error(
+                    "Not a valid entry name",
+                    "not a valid entry name",
+                    from_tag,
+                ))
+            }
+        };
+
+        to.push(from_file_name);
+    }
+
+    std::fs::rename(&from, &to).map_err(|e| {
```

Comment:
> On Unix, if `from` is a directory, `to` must also be an (empty) directory. If `from` is not a directory, `to` must also be not a directory. In contrast, on Windows, `from` can be anything, but `to` must not be a directory.

The check just above here ensures `to` is a filename, not a directory.

This is different from previously, where we had a "copy to new + remove old" approach. Apparently this was added in https://github.com/nushell/nushell/pull/960, to properly move a file between partitions. From the docs, this is an error:

> `from` and `to` are on separate filesystems.

So we definitely want to make that work. Two things for me to do:

1. Add a comment so that someone else doesn't make the mistake of thinking the copy+delete is unnecessary, and
2. Investigate if we can detect filesystems and avoid the copy+delete when on the same filesystem.

Reply:
:+1: